### PR TITLE
[Feature #17518] PROJECT : new menu for template project

### PR DESCRIPTION
**Pull request by psh-axelor**

### CODE REVIEW

File: axelor-project/src/main/resources/views/ProjectTemplate.xml [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+		</panel-tabs>
+
+		<panel name="actionPanel" sidebar="true">
+			<button name="createProjectBtn" icon="fa-copy" title="Create project from this template" onClick="save,action-project-method-generate-project-view" hidden="true" showIf="!isBusinessProject"/>
+			<button name="createBusinessProjectBtn" icon="fa-copy" title="Create business project from this template" onClick="save,action-project-method-generate-project-view" hidden="true" showIf="isBusinessProject"/>
+		</panel>
+
+		<panel name="characteristicsPanel" title="Characteristics" colSpan="12" sidebar="true" canCollapse="true">
+			<panel name="configPanel" colSpan="12">
+				<field if="__config__.app.isApp('business-project')" if-module="axelor-business-project" name="isProject" colSpan="6" widget="boolean-switch" />
```
Comment (sos-axelor) [inline axelor-project/src/main/resources/views/ProjectTemplate.xml]: No need of sub panel put fields directly inside characteristicsPanel , also move all if and if-module condition to that panel.

File: axelor-business-project/src/main/resources/domains/ProjectTemplate.xml [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://axelor.com/xml/ns/domain-models http://axelor.com/xml/ns/domain-models/domain-models_5.0.xsd">
+
+	<module name="project" package="com.axelor.apps.project.db"/>
+
+	<entity name="ProjectTemplate">
+		<boolean name="teamTaskInvoicing" title="Tasks Invoicing"/>
+		<boolean name="isInvoicingExpenses" title="Invoicing Expenses"/>
+		<boolean name="isInvoicingPurchases" title="Invoicing Purchases"/>
+		<integer name="invoicingType" selection="project.invoicing.type.select" title="Invoicing Type"/>
```
Comment (sos-axelor) [inline axelor-business-project/src/main/resources/domains/ProjectTemplate.xml]: field name should be invoicingTypeSelect

File: axelor-project/src/main/resources/views/TaskTemplate.xml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+		<field name="name"/>
+		<field name="assignedTo"/>
+		<field name="team"/>
+		<field name="delayToStart"/>
+		<field name="duration"/>
+		<field name="totalPlannedHrs"/>
+		<field name="isUniqueTaskForMultipleQuantity"/>
+	</grid>
+
+	<form name="task-template-form" title="Task template" model="com.axelor.apps.project.db.TaskTemplate" width="large" onNew="action-product-task-template-on-new,action-product-task-template-attrs-set-required">
```
Comment (sos-axelor) [inline axelor-project/src/main/resources/views/TaskTemplate.xml]: Remove unwanted onNew actions like 'action-product-task-template-on-new'

File: axelor-project/src/main/java/com/axelor/apps/project/web/ProjectTemplateController.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+    ProjectTemplate projectTemplate = request.getContext().asType(ProjectTemplate.class);
+    AppProject appProject = appProjectService.getAppProject();
+
+    if (appProject.getGenerateProjectSequence() && !projectTemplate.getIsBusinessProject()) {
+      Project project = new Project();
+
+      projectService.createProjectFromTemplate(project, projectTemplate);
+
+      response.setCanClose(true);
```
Comment (sos-axelor) [inline axelor-project/src/main/java/com/axelor/apps/project/web/ProjectTemplateController.java]: Not required

File: axelor-project/src/main/java/com/axelor/apps/project/web/ProjectTemplateController.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+      response.setView(
+          ActionView.define(I18n.get("Project"))
+              .model(Project.class.getName())
+              .add("form", "project-form")
+              .add("grid", "project-grid")
+              .context("_showRecord", String.valueOf(project.getId()))
```
Comment (sos-axelor) [inline axelor-project/src/main/java/com/axelor/apps/project/web/ProjectTemplateController.java]: No needs to convert to string for id

File: axelor-project/src/main/java/com/axelor/apps/project/web/ProjectTemplateController.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+  public void createProjectFromWizard(ActionRequest request, ActionResponse response) {
+
+    LinkedHashMap<String, Object> projectTemplateContext =
+        (LinkedHashMap<String, Object>) request.getContext().get("_projectTemplate");
+    Integer projectTemplateId = (Integer) projectTemplateContext.get("id");
+    ProjectTemplate projectTemplate = projectTemplateRepo.find(projectTemplateId.longValue());
+
+    Project project = new Project();
+
+    if (request.getContext().get("code") != null) {
```
Comment (sos-axelor) [inline axelor-project/src/main/java/com/axelor/apps/project/web/ProjectTemplateController.java]: Store context in new variable and get values from it , its not good to use request.getContext repeatedly

File: axelor-project/src/main/java/com/axelor/apps/project/web/ProjectTemplateController.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+      project.setCode((String) request.getContext().get("code"));
+    }
+
+    if (request.getContext().get("clientPartner") != null) {
+      LinkedHashMap<String, Object> clientPartnerContext =
+          (LinkedHashMap<String, Object>) request.getContext().get("clientPartner");
+      Integer clientPartnerId = (Integer) clientPartnerContext.get("id");
```
Comment (sos-axelor) [inline axelor-project/src/main/java/com/axelor/apps/project/web/ProjectTemplateController.java]: Just get context.get("clientPartner").get("id"), no need of null check. Also convert it to string and use Long.parseLong, as integer can't store big numbers.

File: axelor-project/src/main/java/com/axelor/apps/project/service/ProjectServiceImpl.java [hunk 8]
```diff
@@ -145,4 +157,58 @@ public BigDecimal computeDurationFromChildren(Long projectId) {
     q.setParameter("projectId", projectId);
     return q.getSingleResult();
   }
+
+  @Override
+  @Transactional
+  public void createProjectFromTemplate(Project project, ProjectTemplate projectTemplate) {
```
Comment (sos-axelor) [inline axelor-project/src/main/java/com/axelor/apps/project/service/ProjectServiceImpl.java]: remove project as parameter and add clientPartner and code,  method can be call with null value for those two

File: axelor-project/src/main/java/com/axelor/apps/project/web/ProjectTemplateController.java (hunk 7 above)
Comment (psh-axelor) [reply axelor-project/src/main/java/com/axelor/apps/project/web/ProjectTemplateController.java]: when it's not a business project and auto sequence generate is false then clientPartner field will be hidden and null on the wizard. 
So we have to check for null.

**Pull request merged**